app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTaskRepository.kt

```diff
@@ -103,60 +107,74 @@ class DefaultTaskRepository(
         saveTasksToNetwork()
     }
 
     override suspend fun activateTask(taskId: String) {
-        taskDao.updateCompleted(taskId = taskId, completed = false)
+        localDataSource.updateCompleted(taskId = taskId, completed = false)
         saveTasksToNetwork()
     }
 
     override suspend fun clearCompletedTasks() {
-        taskDao.deleteCompleted()
+        localDataSource.deleteCompleted()
         saveTasksToNetwork()
     }
 
     override suspend fun deleteAllTasks() {
-        taskDao.deleteAll()
+        localDataSource.deleteAll()
         saveTasksToNetwork()
     }
 
     override suspend fun deleteTask(taskId: String) {
-        taskDao.deleteById(taskId)
+        localDataSource.deleteById(taskId)
         saveTasksToNetwork()
     }
 
     /**
-     * The following methods load tasks from, and save tasks to, the network.
-     *
-     * Consider these to be long running operations, hence the need for `withContext` which
-     * can change the coroutine dispatcher so that the caller isn't blocked.
+     * The following methods load tasks from (refresh), and save tasks to, the network.
      *
      * Real apps may want to do a proper sync, rather than the "one-way sync everything" approach
      * below. See https://developer.android.com/topic/architecture/data-layer/offline-first
      * for more efficient and robust synchronisation strategies.
      *
      * Also, in a real app, these operations could be scheduled using WorkManager.
+     *
+     * Note that the refresh operation is a suspend function (forces callers to wait) and the save
+     * operation is not. It returns immediately so callers don't have to wait.
      */
-    private suspend fun loadTasksFromNetwork() {
-        withContext(coroutineDispatcher) {
-            val remoteTasks = tasksNetworkDataSource.loadTasks()
-            taskDao.deleteAll()
-            taskDao.upsertAll(remoteTasks.toLocal())
+
+    /**
+     * Delete everything in the local data source and replace it with everything from the network
+     * data source.
+     *
+     * `withContext` is used here in case the bulk `toLocal` mapping operation is complex.
+     */
+    override suspend fun refresh() {
+        withContext(dispatcher) {
+            val remoteTasks = networkDataSource.loadTasks()
+            localDataSource.deleteAll()
+            localDataSource.upsertAll(remoteTasks.toLocal())
         }
     }
 
-    private suspend fun saveTasksToNetwork() {
-        withContext(coroutineDispatcher) {
-            val localTasks = taskDao.getAll()
-            tasksNetworkDataSource.saveTasks(localTasks.toNetwork())
+    /**
+     * Send the tasks from the local data source to the network data source
+     *
+     * Returns immediately after launching the job. Real apps may want to either suspend here until
+     * the operation is complete, or provide a mechanism for the result to be communicated back
+     * to other layers.
```

Comment:
This is a great example of deferred work, so the comment should explicitly say real apps may want to defer to `WorkManager` here. We implement a basic form deferred work by launching an independent coroutine with no exponential backoff. We also probs want to catch any exceptions for completeness sake.

Reply:
Comment updated.